hydro_cli_maelstrom/src/main.rs

```diff
@@ -0,0 +1,351 @@
+    }
+    // send "start: {}" message
+    let connection_defs_str = serde_json::to_string(&connection_defs)?;
+    let formatted_defns = format!("start: {connection_defs_str}\n");
+    child_stdin.write_all(formatted_defns.as_bytes()).await?;
+    #[cfg(debug_assertions)]
+    println!("sent {}", formatted_defns);
+    Ok(())
+}
+
+/// Demux from stdin to all source ports
+/// type_to_port_name is a map from payload type to port name
+async fn input_handler(
+    source_ports: HashMap<String, ServerPort>,
+    type_to_port_name: HashMap<String, String>,
+) -> Result<()> {
+    let mut source_demux = source_demux(source_ports).await?;
+
+    let mut lines = BufReader::new(stdin()).lines();
+    while let Ok(Some(line)) = lines.next_line().await {
+        // Read the line as a json value (maelstrom payload)
+        let mut v: Value = serde_json::from_str(&line)?;
+
+        let msg_id = Value::Array(vec![v["src"].clone(), v["body"]["msg_id"].clone()]);
+
+        // Get the body of the payload
+        let body = v["body"].as_object_mut().unwrap();
+
+        // Update the msg_id to be the pair [src, msg_id]
+        body.insert("msg_id".into(), msg_id);
+
+        // Get the payload type (like echo, echo_ok, etc)
+        let body_type = body.remove("type").unwrap();
+        let body_type = body_type.as_str().unwrap();
+
+        // Get the target port for that payload type
+        let target_port_name = type_to_port_name.get(body_type).unwrap();
+        let target_port = source_demux.get_mut(target_port_name).unwrap();
+
+        // Send the body string to the target port
+        let body_string = serde_json::to_string(body)?;
+        #[cfg(debug_assertions)]
+        println!("Sending line {}", body_string);
+        target_port.send(Bytes::from(body_string)).await?;
+    }
+    Ok(())
+}
+
+/// Accept a connection on each sink port which wraps outputs in maelstrom payload of specified type
+/// Generated "msg_id"s will be `= output_id (mod output_count)` to ensure no overlap
+async fn output_handler(
+    listener: TcpListener,
+    maelstrom_type: String,
+    node_id: String,
+    output_id: usize,
+    output_count: usize,
+) -> Result<()> {
+    let in_stream = listener.accept().await?.0;
+
+    let mut lines = FramedRead::new(in_stream, LengthDelimitedCodec::new());
+    #[cfg(debug_assertions)]
+    println!("accepted connection");
+
+    // Itializer counter which tracks the next available msg_id
+    let mut msg_id_counter = output_id;
+
+    while let Some(Ok(line)) = lines.next().await {
+        // Transforms output into maelstrom payload
+        // For example:
+        // {"echo":"hello world!","in_reply_to":["n1", 1]}
+        // ->
+        // {"src":"n1","dest":"c1","body":{"echo":"hello world!","msg_id":0,"in_reply_to":1,"type":"echo_ok"}}
+
+        // parse line to string
+        let raw_line: String = bincode::deserialize(&line).unwrap();
```

Comment:
This is a bit funky; we should probably have a special `MaelstromResponse` type that's available to both this helper and the actual Hydroflow program. That way instead of using stringly-typed data we can have proper data types.